File: chapter-11/ios-specific/NativeUIComponent/ios/NativeUIComponent/ButtonViewManager.m
//
//  ButtonViewManager.m
//  NativeUIComponent
//
//  This Objective-C file registers the Swift ViewManager with React Native's bridge.
//  Since React Native's view management system is built in Obj-C, Swift ViewManagers
//  need this companion file.
//
//  RCT_EXTERN_MODULE declares the ViewManager name (must match @objc(ButtonViewManager))
//  RCT_EXPORT_VIEW_PROPERTY declares props that can be set from JavaScript
//

#import <React/RCTViewManager.h>

@interface RCT_EXTERN_MODULE(ButtonViewManager, RCTViewManager)

RCT_EXPORT_VIEW_PROPERTY(buttonText, NSString)
RCT_EXPORT_VIEW_PROPERTY(onChange, RCTBubblingEventBlock)

@end

File: chapter-11/ios-specific/NativeUIComponent/ios/NativeUIComponent/ButtonViewManager.swift
import UIKit

/**
 * ButtonViewManager is a custom native UI component that exposes a UIButton
 * to React Native.
 *
 * It extends RCTViewManager, which means:
 * - view() creates the actual UIKit view
 * - Properties set from JS are received via the companion .m file's RCT_EXPORT_VIEW_PROPERTY
 * - Events are sent back to JS by calling the onChange block (mapped to 'topChange')
 *
 * In JavaScript, this component is used via:
 *   const ButtonView = requireNativeComponent('ButtonView');
 *   <ButtonView buttonText="Hello" onChange={handler} />
 */
@objc(ButtonViewManager)
class ButtonViewManager: RCTViewManager {

  override func view() -> UIView! {
    let button = NativeButton()
    return button
  }

  override static func requiresMainQueueSetup() -> Bool {
    return true
  }
}

/**
 * NativeButton is a UIButton subclass that:
 * - Accepts a buttonText prop (sets the title)
 * - Has an onChange callback block (called when tapped, sends event to JS)
 */
class NativeButton: UIButton {

  @objc var buttonText: String = "" {
    didSet {
      setTitle(buttonText, for: .normal)
    }
  }

  /// This block is called by React Native's event system when we fire it.
  /// RN maps 'onChange' prop on the JS side to this block.
  @objc var onChange: RCTBubblingEventBlock?

  override init(frame: CGRect) {
    super.init(frame: frame)
    setup()
  }

  required init?(coder: NSCoder) {
    super.init(coder: coder)
    setup()
  }

  private func setup() {
    backgroundColor = .systemBlue
    setTitleColor(.white, for: .normal)
    addTarget(self, action: #selector(handleTap), for: .touchUpInside)
  }

  @objc private func handleTap() {
    onChange?(["message": "Button clicked!"])
  }
}
